src/main/java/ch/njol/skript/variables/FlatFileStorage.java

```diff
@@ -412,43 +459,166 @@ public final void saveVariables(final boolean finalSave) {
 			}
 		}
 	}
-	
+
 	/**
 	 * Saves the variables.
 	 * <p>
 	 * This method uses the sorted variables map to save the variables in order.
-	 * 
-	 * @param pw
-	 * @param parent The parent's name with {@link Variable#SEPARATOR} at the end
-	 * @param map
+	 *
+	 * @param pw the print writer to write the CSV lines too.
+	 * @param parent The parent's name with {@link Variable#SEPARATOR} at the end.
+	 * @param map the variables map.
 	 */
 	@SuppressWarnings("unchecked")
 	private void save(PrintWriter pw, String parent, TreeMap<String, Object> map) {
-		outer: for (Entry<String, Object> e : map.entrySet()) {
-			Object val = e.getValue();
-			if (val == null)
-				continue;
-			if (val instanceof TreeMap) {
-				save(pw, parent + e.getKey() + Variable.SEPARATOR, (TreeMap<String, Object>) val);
+		// Iterate over all children
+		outer: for (Entry<String, Object> childEntry : map.entrySet()) {
+			Object childNode = childEntry.getValue();
+			String childKey = childEntry.getKey();
+
+			if (childNode == null)
+				continue; // Leaf node
+
+			if (childNode instanceof TreeMap) {
+				// TreeMap found, recurse
+				save(pw, parent + childKey + Variable.SEPARATOR, (TreeMap<String, Object>) childNode);
 			} else {
-				String name = e.getKey() == null ? parent.substring(0, parent.length() - Variable.SEPARATOR.length()) : parent + e.getKey();
+				// Remove variable separator if needed
+				String name = childKey == null ? parent.substring(0, parent.length() - Variable.SEPARATOR.length()) : parent + childKey;
 
 				try {
-					for (VariablesStorage s : Variables.storages) {
-						if (s.accept(name)) {
-							if (s == this) {
-								SerializedVariable.Value value = Classes.serialize(val);
-								if (value != null)
-									writeCSV(pw, name, value.type, encode(value.data));
+					// Loop over storages to make sure this variable is ours to store
+					for (VariablesStorage storage : Variables.STORAGES) {
+						if (storage.accept(name)) {
+							if (storage == this) {
+								// Serialize the value
+								SerializedVariable.Value serializedValue = Classes.serialize(childNode);
+
+								// Write the CSV line
+								if (serializedValue != null)
+									writeCSV(pw, name, serializedValue.type, encode(serializedValue.data));
 							}
+
 							continue outer;
```

Comment:
```suggestion
							break;
```
Break instead of continue label?